Upload captures

LLKruczek opened the pull request:
[SFDS-30](https://crc-dev.atlassian.net/browse/SFDS-30)

Implemented checksum calculation on the browser.
Implemented two helper functions that calls file_endpoint.py methods.
Implemented file upload (without multiplexer) logic.
The capture creation logic is only a place holder and does not create valid captures yet.

lucaspar commented on gateway/sds_gateway/users/migrations/0010_alter_user_is_approved.py:
file not needed

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -64,6 +73,47 @@
+    """Returns True if the path is a valid file for SDS upload."""
+    disallowed_mimes = [
+        "application/x-msdownload",  # .exe
+        "application/x-msdos-program",  # .com
+        "application/x-msi",  # .msi
+    ]
+    mime_type, _ = mimetypes.guess_type(file_path)
+    reasons = []
+    path_obj = Path(file_path)
+    if mime_type in disallowed_mimes:
+        reasons.append(f"Invalid MIME type: {mime_type}")
+    if not path_obj.is_file():
+        reasons.append("Not a file")
+    if path_obj.stat().st_size == 0:
+        reasons.append("Empty file")
+    return len(reasons) == 0, reasons
+
+
+def get_valid_files(root_dir: Path):
+    """
+    Recursively yields (relative_path, file_path) for valid files under root_dir.
+    """
+    for file_path in root_dir.rglob("*"):
+        if not file_path.is_file():
+            continue
+        is_valid, reasons = is_valid_file(str(file_path))
+        if is_valid:
+            rel_path = str(file_path.relative_to(root_dir))
+            yield rel_path, file_path
+
+
+def compute_blake3_checksum(file_path):
+    # Use hashlib.blake2b with digest_size=32 to match 64 hex chars
+    h = hashlib.blake2b(digest_size=32)
+    with file_path.open("rb") as f:
+        for chunk in iter(lambda: f.read(8192), b""):
+            h.update(chunk)
+    return h.hexdigest()
+
+
```
these are not called anywhere; let's make them part of a future PR if they'll be used in the future.

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -1609,6 +1659,173 @@ def _serve_file_download(self, zip_uuid: str, user) -> HttpResponse:
+        saved_files = []
+        errors = []
+        skipped_files = []
+        metadata_only_files = []
```
`metadata_only_files` is always empty

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -1609,6 +1659,173 @@ def _serve_file_download(self, zip_uuid: str, user) -> HttpResponse:
+        metadata_only_files = []
+
+        for f, rel_path in zip(files, relative_paths, strict=True):
+            if "/" in rel_path:
+                directory, filename = rel_path.rsplit("/", 1)
+                directory = (
+                    "/" + directory if not directory.startswith("/") else directory
+                )
+            else:
+                directory = "/"
+                filename = rel_path
```
use [`pathlib`](https://docs.python.org/3/library/pathlib.html#methods-and-properties) methods and properties: it's less error-prone.

lucaspar commented on gateway/sds_gateway/api_methods/helpers/file_helpers.py:
```diff
@@ -0,0 +1,74 @@
+
+
+def upload_file_helper_simple(request, file_data):
+    """
+    Upload a single file using FileViewSet.create.
+    file_data should contain all required fields: name, directory, file,
+    media_type, etc.
+    Returns ([response], []) for success, ([], [error]) for error, and handles
+    409 as a warning.
+    """
```
[nit] try to follow this style guide for docstrings; you'll see it in other methods too

https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

lucaspar commented on gateway/sds_gateway/api_methods/helpers/file_helpers.py:
```diff
@@ -0,0 +1,74 @@
+            # Already exists, treat as warning
+            return [], [getattr(response, "data", str(response))]
+        return [], [getattr(response, "data", str(response))]
+    except Exception as e:  # noqa: BLE001
+        return [], [str(e)]
+
```
too broad, we should catch a specific exception, if any. This might mask problems with the code syntax, the logic, and other things we definitely want to raise instead of returning as an error

lucaspar commented on gateway/sds_gateway/api_methods/helpers/file_helpers.py:
```diff
@@ -0,0 +1,74 @@
+    view.kwargs = {}
+    try:
+        response = view.create(drf_request)
+        if (
+            hasattr(response, "status_code")
+            and status.HTTP_200_OK
+            <= response.status_code
+            < status.HTTP_300_MULTIPLE_CHOICES
+        ):
+            return [response], []
+        if (
+            hasattr(response, "status_code")
+            and response.status_code == status.HTTP_409_CONFLICT
+        ):
+            # Already exists, treat as warning
+            return [], [getattr(response, "data", str(response))]
```
try this

```py
from http import HTTPStatus
status = HTTPStatus(status_code)
status.is_success       # 2xx
status.is_client_error  # 4xx
status.is_server_error	# 5xx
```

lucaspar replied:
5xx response is not being handled

lucaspar commented on gateway/sds_gateway/templates/users/file_list.html:
```diff
@@ -323,5 +390,124 @@ <h2 class="accordion-header" id="headingCategory">
+        });
+    </script>
+    <!-- Load the WASM library first -->
+    <script src="https://cdn.jsdelivr.net/npm/hash-wasm@4.12.0/dist/blake3.umd.min.js"></script>
+    <script>
+        document.addEventListener('DOMContentLoaded', function() {
+            var modal = document.getElementById('uploadCaptureModal');
+            if (!modal) {
+                console.warn('uploadCaptureModal not found');
+                return;
+            }
+            modal.addEventListener('shown.bs.modal', function() {
+                var fileInput = document.getElementById('captureFileInput');
+                if (!fileInput) {
+                    console.warn('captureFileInput not found');
+                    return;
+                }
+                // Remove any previous handler to avoid duplicates
+                fileInput.removeEventListener('change', window._blake3CaptureHandler);
+
+                window._blake3CaptureHandler = async function(event) {
+                    console.log('File input changed!');
+                    const files = event.target.files;
+                    if (!files || files.length === 0) {
+                        console.log("No files selected.");
+                        return;
+                    }
+                    for (const file of files) {
+                        const buffer = await file.arrayBuffer();
+                        const hasher = await hashwasm.createBLAKE3(); // async!
+                        hasher.init();
+                        hasher.update(new Uint8Array(buffer));
+                        const hashHex = hasher.digest('hex');
+                        console.log(`BLAKE3 hash for ${file.name}:`, hashHex);
+                    }
+                };
+                fileInput.addEventListener('change', window._blake3CaptureHandler);
+            });
+        });
     </script>
```
remove or comment it out for a future PR

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -1609,6 +1659,173 @@ def _serve_file_download(self, zip_uuid: str, user) -> HttpResponse:
+                        f"Failed to create capture for channel {channel}: "
+                        f"{serializer.errors}"
+                    )
+            except Exception as exc:
+                logger.exception("Failed to create capture for channel %s", channel)
+                errors.append(
+                    f"Failed to create capture for channel {channel}: {exc!s}"
+                )
+                continue
```
too broad

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -1609,6 +1659,173 @@ def _serve_file_download(self, zip_uuid: str, user) -> HttpResponse:
+                        "DEBUG: Created capture with uuid: %s",
+                        capture.uuid,
+                    )
+                    created_captures.append(
+                        {
+                            "uuid": str(capture.uuid)
+                            if hasattr(capture, "uuid")
+                            else None,
+                            "channel": channel,
+                            "top_level_dir": str(top_level_dir),
+                        }
+                    )
```
can this be simplified to `created_captures.append(capture)`?

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -1852,11 +1805,19 @@ def _create_captures(self, request, channels, relative_paths):
-                errors.append(
-                    f"Failed to create capture for channel {channel}: {exc!s}"
+                logger.exception(
+                    "Unexpected error creating capture for channel %s", channel
```
too general exception

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -1806,3 +1815,567 @@ def post(
+user_upload_files_view = UploadFilesView.as_view()
+
+
+# TODO: Use this view when implementing the file upload mode multiplexer.
```
I see you included checksum code, is this still a work in progress?

lucaspar commented on gateway/sds_gateway/api_methods/helpers/file_helpers.py:
```diff
@@ -0,0 +1,154 @@
+        return responses, errors  # noqa: TRY300
+    except (ValueError, TypeError, AttributeError, KeyError) as e:
+        return [], [f"Data validation error: {e}"]
+    except Exception as e:  # noqa: BLE001
+        return [], [f"Unexpected error: {e}"]
```
leave uncaught

lucaspar commented on gateway/sds_gateway/api_methods/helpers/file_helpers.py:
```diff
@@ -0,0 +1,154 @@
+        return responses, errors  # noqa: TRY300
+    except (ValueError, TypeError, AttributeError, KeyError) as e:
+        return [], [f"Data validation error: {e}"]
+    except Exception as e:  # noqa: BLE001
+        return [], [f"Unexpected error: {e}"]
```
leave uncaught

lucaspar commented on gateway/sds_gateway/api_methods/helpers/file_helpers.py:
```diff
@@ -0,0 +1,154 @@
+                # Handle any other status codes
+                errors.append(response_data)
+
+        return responses, errors  # noqa: TRY300
```
return out of the try block

https://docs.astral.sh/ruff/rules/try-consider-else/

lucaspar commented on gateway/sds_gateway/api_methods/helpers/file_helpers.py:
```diff
@@ -0,0 +1,154 @@
+                # Handle any other status codes
+                errors.append(response_data)
+
+        return responses, errors  # noqa: TRY300
```
return out of the try block

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -1806,3 +1815,567 @@ def post(
+                channel,
+                error_msg,
+            )
+            return (  # noqa: TRY300
+                None,
+                f"Failed to create capture for channel {channel}: {error_msg}",
+            )
```
return out of the try block

https://docs.astral.sh/ruff/rules/try-consider-else/

LLKruczek commented on gateway/config/settings/base.py:
```diff
@@ -498,3 +498,16 @@ def __get_random_token(length: int) -> str:
     "SDS_NEW_USERS_APPROVED_ON_CREATION",
     default=False,
 )
+
+# File upload limits
+# ------------------------------------------------------------------------------
+# Maximum number of files that can be uploaded at once
+DATA_UPLOAD_MAX_NUMBER_FILES: int = env.int(
+    "DATA_UPLOAD_MAX_NUMBER_FILES", default=1000
+)
+
+# Maximum memory size for file uploads (default: 2.5MB, increased to 100MB)
+DATA_UPLOAD_MAX_MEMORY_SIZE: int = env.int(
+    "DATA_UPLOAD_MAX_MEMORY_SIZE",
+    default=104857600,  # 100MB
+)
```
Yes the maximum amount of files are increased because radiohound folder usually have hundreds of files that exceeded the previous constraint. The memory increase is not needed now (after uploading by chunks). Feel free to advise on a good number to set memory limit to. Thanks!

LLKruczek commented on gateway/sds_gateway/api_methods/views/file_endpoints.py:
```diff
@@ -78,9 +98,15 @@ class FileViewSet(ViewSet):
     def create(self, request: Request) -> Response:
         """Uploads a file to the server."""
 
+        # Check for client disconnection
+        if self._is_client_disconnected(request):
+            return Response(
+                {"detail": "Client disconnected"},
+                status=HTTP_499_CLIENT_CLOSED_REQUEST,
+            )
+
```
This is part of orphaned disconnection detection code on the backend that didn't run... I now moved disconnection control to the frontend. If connection drops the backend will just finish processing the current chunk and then stop. Let me know if you would like to have some disconnection detection on the backend too.

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -1800,3 +1819,562 @@ def get(self, request, *args, **kwargs) -> JsonResponse:
+                "sum_blake3": checksum,
+            }
+
+            # Call the helper function
+            response = check_file_contents_exist_helper(request, check_data)
+
+            # Extract the response data
+            if hasattr(response, "data"):
+                response_data = response.data
+            else:
+                response_data = str(response)
+
+            # Return the result
+            http_ok = 200
+            return JsonResponse(
+                {
+                    "exists": response.status_code == http_ok,
+                    "status_code": response.status_code,
+                    "data": response_data,
+                }
+            )
```
This interprets a successful response as "the file exists" (`"exists": response.status_code == http_ok`), but `CheckFileContentsExistView.post` (the method requested by the helper) will also succeed if the file doesn't exist.

See `CheckFileContentsExistView.post`'s docstring to know what to look for in the response payload. Please reach out if this comment is not clear.

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -1800,3 +1819,562 @@ def get(self, request, *args, **kwargs) -> JsonResponse:
+                total_chunks,
+            )
+
+        # Link all uploaded files to all created captures
+        # (only if files were actually uploaded)
+        if saved_files and created_captures:
+            # Get File objects by uuid
+            file_objs = File.objects.filter(uuid__in=file_uuids)
+            capture_uuids = [c["uuid"] for c in created_captures if c.get("uuid")]
+            captures = Capture.objects.filter(uuid__in=capture_uuids)
+            for capture in captures:
+                file_objs.update(capture=capture)
+        elif errors and not all_files_empty:
+            logger.error(
+                "File upload errors occurred, skipping capture creation. Errors: %s",
+                errors,
+            )
```
Files are already linked to captures during ingestion, we're not doing that manually, nor allowing arbitrary files to be linked - that'd be the role of a dataset.

We can probably drop this selected block and adjust the lines below.

lucaspar commented on gateway/sds_gateway/api_methods/utils/metadata_schemas.py:
```diff
@@ -406,14 +404,14 @@ def get_mapping_by_capture_type(
     """Infer the index name for a given capture."""
     # Populate index_name based on capture type
     match capture_type:
-        case CaptureType.DigitalRF:
-            return f"captures-{CaptureType.DigitalRF}"
-        case CaptureType.RadioHound:
-            return f"captures-{CaptureType.RadioHound}"
+        case "drf":
+            return "captures-drf"
+        case "rh":
+            return "captures-rh"
```
ah I see it now. Moving CaptureType will require changing at least 14 other places; do that local import and add a comment that it's local to avoid the circular import.

LLKruczek commented on gateway/sds_gateway/api_methods/views/capture_endpoints.py:
```diff
@@ -273,36 +273,172 @@ def create(self, request: Request) -> Response:  # noqa: PLR0911
+            capture_type, requested_top_level_dir, requester, drf_channel, rh_scan_group
+        )
+        if response is not None:
+            return response
```
I just pushed a new change that refactored ingest_capture() to avoid writing empty captures in case of channel error. Let me know if this works!

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -2187,27 +2139,38 @@ def _handle_capture_creation(
+                all_relative_paths, all_relative_paths
+            )
+
+            # Determine capture type enum
+            if capture_type == "rh":
+                capture_type_enum = CaptureType.RadioHound
+            else:
+                capture_type_enum = CaptureType.DigitalRF
+
+            # Prepare base capture data
+            capture_data = {
+                "capture_type": capture_type_enum,
```
just inject `capture_type` directly into the `capture_data` dict

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -1858,34 +1863,21 @@ def _handle_file_uploads(self, request, files, relative_paths):
-                            "status": response.status_code,
-                        }
-                    )
+                    saved_files.append({"uploaded": True})
```
It seems you're not tracking the file UUIDs anymore, which is fine _if you're not using this information_. But then, maybe `saved_files` can just be a `num_saved_files` counter instead of a list?

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -1913,7 +1905,9 @@ def _create_capture_with_endpoint_helper(
-                    return capture_data, None
+                    # Extract only the UUID since that's all we use
+                    capture_uuid = capture_data.get("uuid")
+                    return {"uuid": capture_uuid}, None
```
likewise, just return the UUID instead of a dict

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -2054,24 +2059,27 @@ def file_upload_status_mux(
             # (they might already exist)
             return "success" if has_required_fields else "error"
-        if saved_files and len(saved_files) == len(files) and not errors:
+        if (
+            saved_files
+            and len(saved_files) == len(upload_chunk_files)
+            and not file_errors
+        ):
```
this can be simplified if `saved_files` becomes a counter integer

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -2054,24 +2059,27 @@ def file_upload_status_mux(
-        errors,
-        capture_errors,
+        file_upload_status: str,
+        saved_files: list[dict[str, Any]],
```
check `saved_files` usages in this function, consider dropping this argument

lucaspar commented on gateway/sds_gateway/api_methods/utils/metadata_schemas.py:
```diff
@@ -406,14 +424,28 @@ def get_mapping_by_capture_type(
+    # Local import to avoid circular dependency
+    from sds_gateway.api_methods.models import CaptureType
+
+    # Handle both string and enum inputs
+    if isinstance(capture_type, str):
+        # Convert string to enum for comparison
+        if capture_type == "drf":
+            return "captures-drf"
+        if capture_type == "rh":
+            return "captures-rh"
+        msg = f"Invalid capture type string: {capture_type}"
+        log.error(msg)
+        raise ValueError(msg)
```
not needed, `str` matches fine against `StrEnum`

lucaspar commented on gateway/sds_gateway/api_methods/utils/metadata_schemas.py:
```diff
@@ -387,9 +383,31 @@
+            }
+        )
+
+    # Handle string inputs by converting to enum
+    if isinstance(capture_type, str):
+        if capture_type == "drf":
+            capture_type = CaptureType.DigitalRF
+        elif capture_type == "rh":
+            capture_type = CaptureType.RadioHound
+        else:
+            msg = f"Invalid capture type string: {capture_type}"
+            log.error(msg)
+            raise ValueError(msg)
```
not needed, just use `capture_type: CaptureType` directly as a dict key of `capture_index_mapping_by_type` below

lucaspar commented on gateway/sds_gateway/api_methods/utils/metadata_schemas.py:
```diff
@@ -406,14 +424,28 @@ def get_mapping_by_capture_type(
+    # Local import to avoid circular dependency
+    from sds_gateway.api_methods.models import CaptureType
+
+    # Handle both string and enum inputs
+    if isinstance(capture_type, str):
+        # Convert string to enum for comparison
+        if capture_type == "drf":
+            return "captures-drf"
+        if capture_type == "rh":
+            return "captures-rh"
+        msg = f"Invalid capture type string: {capture_type}"
+        log.error(msg)
+        raise ValueError(msg)
+
```
not needed, `str` matches fine against `StrEnum` below

lucaspar commented on gateway/sds_gateway/api_methods/utils/metadata_schemas.py:
```diff
@@ -406,14 +424,28 @@ def get_mapping_by_capture_type(
 
 
-def infer_index_name(capture_type: CaptureType) -> str:
+def infer_index_name(capture_type: Any) -> str:
```
this fn is fine to accept `str`, but we should annotate it as `capture_type: CaptureType`

lucaspar commented on gateway/sds_gateway/api_methods/views/capture_endpoints.py:
```diff
@@ -273,39 +274,84 @@ def create(self, request: Request) -> Response:  # noqa: PLR0911
+                    data=request.data.copy(),
+                    context={"request_user": request.user},
+                )
+                post_serializer.is_valid()  # Already validated above
```
I think this comment is not true anymore

LLKruczek commented:
Changes in commit [4f15ed6](https://github.com/spectrumx/sds-code/pull/160/commits/4f15ed6e68ce2112641a8dd5738fd90b040d7d77)

1. Added error catching to the backend to return JSON response in case of Django default error.
2. Changed file chunking to group files until they are 50MB, and upload larger files individually. Disallow files larger than 512MB.
3. MInor enhancement of error message display (such as prompting the user to try again or to use the SDK instead).

lucaspar commented on gateway/sds_gateway/users/views.py:
```diff
@@ -2176,124 +2187,163 @@ def _process_capture_creation(
+                },
+                status=400,
+            )
+        except (ConnectionError, TimeoutError) as e:
+            logger.exception("Network error in UploadCaptureView.post")
+            return JsonResponse(
+                {
+                    "success": False,
+                    "error": "Network connection error",
+                    "error_code": "NETWORK_ERROR",
+                    "message": f"Network error: {e!s}",
+                },
+                status=503,
+            )
+        except Exception as e:
+            logger.exception("Unexpected error in UploadCaptureView.post")
+            return JsonResponse(
+                {
+                    "success": False,
+                    "error": "Internal server error",
+                    "error_code": "UNKNOWN_ERROR",
+                    "message": f"{e!s}",
+                },
+                status=500,
+            )
```
are these `"error_code"` fields being used?
